File: xrecycleviewlib/src/main/java/com/ashlikun/xrecycleview/RecyclerViewExtend.java
package com.ashlikun.xrecycleview;

import android.content.Context;
import android.content.res.TypedArray;
import android.support.v7.widget.GridLayoutManager;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.support.v7.widget.StaggeredGridLayoutManager;
import android.util.AttributeSet;
import android.util.Log;
import android.view.View;
import android.view.ViewGroup;

import java.lang.reflect.Field;
import java.util.ArrayList;
import java.util.List;

/**
 * 作者　　: 李坤
 * 创建时间: 10:12 Administrator
 * 邮箱　　：496546144@qq.com
 * <p>
 * 功能介绍：扩展的RecyclerView
 * 1：带头部与底部的RecycleView
 * 2：上下加载更多
 * 3：和可以设置最大高度
 */

public class RecyclerViewExtend extends RecyclerView {

    private static final String HEADERSIZE = "headerSize";
    private static final String FOOTERSIZE = "footerSize";
    private static final int TYPE_HEADER = -900004;
    private static final int TYPE_REFRESH_HEADER = -900005;
    private static final int TYPE_NORMAL = 0;
    private static final int TYPE_REFRESH_FOOTER = -900003;
    private static final int TYPE_FOOTER = -900002;
    /**
     * 第一个加载的布局
     */
    protected ArrayList<View> mHeaderViews = new ArrayList<>();
    /**
     * 最后一个为加载的布局
     */
    protected ArrayList<View> mFootViews = new ArrayList<>();
    private Adapter mAdapter;
    private WrapAdapter mWrapAdapter;
    /**
     * 最大高度
     */
    protected float maxHeight = 0;
    /**
     * 最大比例,相对于宽度
     */
    protected float maxRatio = 0;

    public boolean isHeader(ViewHolder viewHolder) {
        return viewHolder.getItemViewType() == TYPE_REFRESH_HEADER || viewHolder.getItemViewType() == TYPE_HEADER;
    }

    public boolean isFooter(ViewHolder viewHolder) {
        return viewHolder.getItemViewType() == TYPE_REFRESH_FOOTER || viewHolder.getItemViewType() == TYPE_FOOTER;
    }

    public RecyclerViewExtend(Context context) {
        this(context, null);
    }

    public RecyclerViewExtend(Context context, AttributeSet attrs) {
        this(context, attrs, 0);
    }

    public RecyclerViewExtend(Context context, AttributeSet attrs, int defStyle) {
        super(context, attrs, defStyle);
        initView(context, attrs);
    }


    private void initView(Context context, AttributeSet attrs) {
        TypedArray a = context.obtainStyledAttributes(attrs,
                R.styleable.RecyclerViewExtend);
        maxRatio = a.getFloat(R.styleable.RecyclerViewExtend_rv_heightRatio, 0);
        maxHeight = a.getDimension(R.styleable.RecyclerViewExtend_rv_heightDimen, 0);
        a.recycle();
    }

    public int getDataSize() {
        return mAdapter.getItemCount();
    }

    @Override
    protected void onMeasure(int widthMeasureSpec, int heightMeasureSpec) {
        if (maxHeight <= 0 && maxRatio <= 0) {
            super.onMeasure(widthMeasureSpec, heightMeasureSpec);
            return;
        }
        int widthSize = MeasureSpec.getSize(widthMeasureSpec);
        if (maxHeight <= 0) {
            maxHeight = maxRatio * widthSize;
        } else {
            maxHeight = Math.min(maxHeight, maxRatio * widthSize);
        }
        if (maxHeight <= 0) {
            super.onMeasure(widthMeasureSpec, heightMeasureSpec);
            return;
        }
        int heightMode = MeasureSpec.getMode(heightMeasureSpec);
        int heightSize = MeasureSpec.getSize(heightMeasureSpec);

        if (heightMode == MeasureSpec.EXACTLY) {
            heightSize = heightSize <= maxHeight ? heightSize
                    : (int) maxHeight;
        }

        if (heightMode == MeasureSpec.UNSPECIFIED) {
            heightSize = heightSize <= maxHeight ? heightSize
                    : (int) maxHeight;
        }
        if (heightMode == MeasureSpec.AT_MOST) {
            heightSize = heightSize <= maxHeight ? heightSize
                    : (int) maxHeight;
        }
        int maxHeightMeasureSpec = MeasureSpec.makeMeasureSpec(heightSize,
                heightMode);
        super.onMeasure(widthMeasureSpec, maxHeightMeasureSpec);
    }

    @Override
    public void setAdapter(Adapter adapter) {
        mAdapter = adapter;
        mWrapAdapter = new WrapAdapter();
        super.setAdapter(mWrapAdapter);
        try {
            mAdapter.registerAdapterDataObserver(mDataObserver);
        } catch (IllegalStateException e) {

        }
        mWrapAdapter.notifyDataSetChanged();
        setHeaderSize();
        setFooterSize();
    }

    public void addHeaderView(View view) {
        mHeaderViews.add(view);
        setHeaderSize();
    }

    public void addHeaderView(int index, View view) {
        mHeaderViews.add(index, view);
        setHeaderSize();
    }

    public View getHeaderView(int index) {
        if (mHeaderViews.size() > index && index >= 0) {
            return mHeaderViews.get(index);
        }
        return null;
    }

    public void removeHeaderView(View view) {
        mHeaderViews.remove(view);
        setHeaderSize();
    }

    public void removeFootView(View view) {
        mFootViews.remove(view);
        setFooterSize();
    }

    public void addFootView(final View view) {
        mFootViews.add(view);
        setFooterSize();
    }

    public void addFootView(int index, final View view) {
        mFootViews.add(index, view);
        setFooterSize();
    }

    public View getFootView(int index) {
        if (mFootViews.size() > index && index >= 0) {
            return mFootViews.get(index);
        }
        return null;
    }

    protected void setFooterSize() {
        if (mAdapter == null) {
            return;
        }
        //应为CommonAdapter为抽象类
        Class cls = getCommonAdapterClass(mAdapter.getClass());
        if (cls == null) {
            return;
        }
        try {
            Field field = cls.getDeclaredField(FOOTERSIZE);
            field.setAccessible(true);
            try {
                field.set(mAdapter, getFootViewSize());
            } catch (IllegalAccessException e) {
                Log.w("setFooterSize", "adapter设置" + FOOTERSIZE + "字段失败");
            }
        } catch (NoSuchFieldException e) {
            Log.w("setFooterSize", "adapter没有" + FOOTERSIZE + "字段");
        }

    }

    private void setHeaderSize() {
        if (mAdapter == null) {
            return;
        }
        Class cls = getCommonAdapterClass(mAdapter.getClass());//应为CommonAdapter为抽象类
        if (cls == null) {
            return;
        }
        try {
            Field field = cls.getDeclaredField(HEADERSIZE);
            field.setAccessible(true);
            try {
                field.set(mAdapter, getHeaderViewSize());
            } catch (IllegalAccessException e) {
                Log.w("setHeaderSize", "adapter设置" + HEADERSIZE + "字段失败");
            }
        } catch (NoSuchFieldException e) {
            Log.w("setHeaderSize", "adapter没有" + HEADERSIZE + "字段");
        }

    }

    private Class<? extends Adapter> getCommonAdapterClass(Class cls) {
        if (cls.isAssignableFrom(Adapter.class) || cls.getSuperclass() == null) {
            return null;
        } else {
            try {
                cls.getDeclaredField(HEADERSIZE);
                cls.getDeclaredField(FOOTERSIZE);
                return cls;
            } catch (NoSuchFieldException e) {
                return getCommonAdapterClass(cls.getSuperclass());
            }
        }
    }


    public void scrollToPositionWithOffset(int postion) {
        LayoutManager layoutManager = getLayoutManager();
        if (layoutManager instanceof LinearLayoutManager) {
            ((LinearLayoutManager) getLayoutManager()).scrollToPositionWithOffset(postion, 0);
        } else if (layoutManager instanceof GridLayoutManager) {
            ((GridLayoutManager) getLayoutManager()).scrollToPositionWithOffset(postion, 0);
        } else if (layoutManager instanceof StaggeredGridLayoutManager) {
            ((StaggeredGridLayoutManager) getLayoutManager()).scrollToPositionWithOffset(postion, 0);
        }
    }


    public int getHeaderViewSize() {
        return mHeaderViews.size();
    }

    public int getFootViewSize() {
        return mFootViews.size();
    }

    public int getFootViewHeight() {
        int height = 0;
        for (View v : mFootViews) {
            height += v.getHeight();
        }
        return height;
    }

    public int getHeaderViewHeight() {
        int height = 0;
        for (View v : mHeaderViews) {
            height += v.getHeight();
        }
        return height;
    }


    private int findMax(int[] lastPositions) {
        int max = lastPositions[0];
        for (int value : lastPositions) {
            if (value > max) {
                max = value;
            }
        }
        return max;
    }

    /**
     * 当adapter不是调用全部刷新，并且有动画的时候
     */
    protected void onAdapterItemAnimChang() {

    }

    private final AdapterDataObserver mDataObserver = new AdapterDataObserver() {
        @Override
        public void onChanged() {
            if (mAdapter != null) {
                mWrapAdapter.notifyDataSetChanged();
            }
        }

        @Override
        public void onItemRangeInserted(int positionStart, int itemCount) {
            mWrapAdapter.notifyItemRangeInserted(getHeaderViewSize() + positionStart, itemCount);
            if (getItemAnimator() != null) {
                onAdapterItemAnimChang();
            }
        }

        @Override
        public void onItemRangeChanged(int positionStart, int itemCount) {
            mWrapAdapter.notifyItemRangeChanged(getHeaderViewSize() + positionStart, itemCount);
        }

        @Override
        public void onItemRangeChanged(int positionStart, int itemCount, Object payload) {
            mWrapAdapter.notifyItemRangeChanged(getHeaderViewSize() + positionStart, itemCount, payload);
        }

        @Override
        public void onItemRangeRemoved(int positionStart, int itemCount) {
            mWrapAdapter.notifyItemRangeRemoved(getHeaderViewSize() + positionStart, itemCount);
            if (getItemAnimator() != null) {
                onAdapterItemAnimChang();
            }
        }

        @Override
        public void onItemRangeMoved(int fromPosition, int toPosition, int itemCount) {
            mWrapAdapter.notifyItemMoved(getHeaderViewSize() + fromPosition, getHeaderViewSize() + toPosition);
            if (getItemAnimator() != null) {
                onAdapterItemAnimChang();
            }
        }
    };


    private class WrapAdapter extends RecyclerView.Adapter<ViewHolder> {
        private int headerPosition = 0;
        private int footerPosition = 0;

        public WrapAdapter() {
        }


        public boolean isHeader(int position) {
            return position >= 0 && position < mHeaderViews.size();
        }

        public boolean isHeaderLoad(int position) {
            return mHeaderViews.size() > 0 && position == 0 && mHeaderViews.get(0) instanceof LoadView;
        }

        public boolean isFooter(int position) {
            return position < getItemCount() && position >= getItemCount() - mFootViews.size();
        }

        public boolean isFooterLoad(int position) {
            return mFootViews.size() > 0 && position == getItemCount() - 1 && mFootViews.get(mFootViews.size() - 1) instanceof LoadView;
        }

        public int getHeadersCount() {
            return mHeaderViews.size();
        }

        public int getFootersCount() {
            return mFootViews.size();
        }

        @Override
        public void onAttachedToRecyclerView(RecyclerView recyclerView) {
            mAdapter.onAttachedToRecyclerView(recyclerView);
            LayoutManager manager = recyclerView.getLayoutManager();
            if (manager instanceof GridLayoutManager) {
                //可能之前已经有自定义的了,这里不能直接替换
                final GridLayoutManager gridManager = ((GridLayoutManager) manager);
                final GridLayoutManager.SpanSizeLookup old = gridManager.getSpanSizeLookup();
                gridManager.setSpanSizeLookup(new GridLayoutManager.SpanSizeLookup() {
                    @Override
                    public int getSpanSize(int position) {
                        if (isHeader(position) || isFooter(position)) {
                            //占满全行
                            return gridManager.getSpanCount();
                        } else {
                            if (old != null) {
                                //使用之前设置过的 并且positions是去除头部后的位置
                                return old.getSpanSize(position - getHeadersCount());
                            } else {
                                return 1;
                            }
                        }
                    }
                });
                gridManager.setSpanCount(gridManager.getSpanCount());
            }
        }

        @Override
        public void onDetachedFromRecyclerView(RecyclerView recyclerView) {
            mAdapter.onDetachedFromRecyclerView(recyclerView);
        }

        @Override
        public void onViewAttachedToWindow(ViewHolder holder) {
            try {
                if (RecyclerViewExtend.this.isHeader(holder) || RecyclerViewExtend.this.isFooter(holder)) {
                    ViewGroup.LayoutParams lp = holder.itemView.getLayoutParams();
                    if (lp != null
                            && lp instanceof StaggeredGridLayoutManager.LayoutParams
                            && (isHeader(holder.getLayoutPosition()) || isFooter(holder.getLayoutPosition()))) {
                        StaggeredGridLayoutManager.LayoutParams p = (StaggeredGridLayoutManager.LayoutParams) lp;
                        //StaggeredGridLayoutManager 只能这样设置一整行
                        p.setFullSpan(true);
                    }
                } else {
                    mAdapter.onViewAttachedToWindow(holder);
                }
            } catch (Exception e) {
            }
        }

        @Override
        public void onViewDetachedFromWindow(ViewHolder holder) {
            try {
                if (RecyclerViewExtend.this.isHeader(holder) || RecyclerViewExtend.this.isFooter(holder)) {

                } else {
                    mAdapter.onViewDetachedFromWindow(holder);
                }
            } catch (Exception e) {

            }
        }

        @Override
        public ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
            if (viewType == TYPE_REFRESH_FOOTER) {
                return new ViewHolder(getFootView(mFootViews.size() - 1)) {
                };
            } else if (viewType == TYPE_REFRESH_HEADER) {
                return new ViewHolder(getHeaderView(0)) {
                };
            } else if (viewType == TYPE_HEADER) {
                if (headerPosition >= mHeaderViews.size()) {
                    headerPosition = 0;
                }
                return new ViewHolder(mHeaderViews.get(headerPosition++)) {
                };
            } else if (viewType == TYPE_FOOTER) {
                if (footerPosition >= mFootViews.size()) {
                    footerPosition = 0;
                }
                return new ViewHolder(mFootViews.get(footerPosition++)) {
                };
            }
            return mAdapter.onCreateViewHolder(parent, viewType);
        }

        @Override
        public void onBindViewHolder(ViewHolder holder, int position, List<Object> payloads) {
            super.onBindViewHolder(holder, position, payloads);
            if (isFooterLoad(position) || isHeaderLoad(position)) {
                return;
            }
            if (isHeader(position)) {
                return;
            }
            if (isFooter(position)) {
                return;
            }
            int adjPosition = position - getHeadersCount();
            if (mAdapter != null) {
                if (adjPosition < mAdapter.getItemCount()) {
                    mAdapter.onBindViewHolder(holder, adjPosition, payloads);
                    return;
                }
            }
        }

        @Override
        public void onBindViewHolder(ViewHolder holder, int position) {

        }

        @Override
        public int getItemCount() {
            if (mAdapter != null) {
                return getHeadersCount() + getFootersCount() + mAdapter.getItemCount();
            } else {
                return getHeadersCount() + getFootersCount();
            }
        }

        @Override
        public int getItemViewType(int position) {
            if (isFooterLoad(position)) {
                return TYPE_REFRESH_FOOTER;
            }
            if (isHeaderLoad(position)) {
                return TYPE_REFRESH_HEADER;
            }
            if (isHeader(position)) {
                return TYPE_HEADER;
            }
            if (isFooter(position)) {
                return TYPE_FOOTER;
            }
            int adjPosition = position - getHeadersCount();
            if (mAdapter != null) {
                int adapterCount = mAdapter.getItemCount();
                if (adjPosition < adapterCount) {
                    return mAdapter.getItemViewType(adjPosition);
                }
            }
            return TYPE_NORMAL;
        }

        @Override
        public long getItemId(int position) {
            if (mAdapter != null && position >= getHeadersCount()) {
                int adjPosition = position - getHeadersCount();
                int adapterCount = mAdapter.getItemCount();
                if (adjPosition < adapterCount) {
                    return mAdapter.getItemId(adjPosition);
                }
            }
            return -1;
        }

        @Override
        public void unregisterAdapterDataObserver(AdapterDataObserver observer) {

            if (mAdapter != null) {
                try {
                    mAdapter.unregisterAdapterDataObserver(observer);
                } catch (IllegalStateException e) {

                }
            }
        }

        @Override
        public void registerAdapterDataObserver(AdapterDataObserver observer) {
            if (mAdapter != null) {
                try {
                    mAdapter.registerAdapterDataObserver(observer);
                } catch (IllegalStateException e) {

                }
            }
        }
    }

    public void setMaxHeight(float maxHeight) {
        this.maxHeight = maxHeight;
        requestLayout();
    }

    public void setMaxRatio(float maxRatio) {
        this.maxRatio = maxRatio;
        requestLayout();
    }
}

File: xrecycleviewlib/src/main/java/com/ashlikun/xrecycleview/RecyclerViewExtend.kt
package com.ashlikun.xrecycleview

import android.content.Context
import android.util.AttributeSet
import android.view.MotionEvent
import android.view.View
import android.view.ViewGroup
import androidx.recyclerview.widget.GridLayoutManager
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import androidx.recyclerview.widget.StaggeredGridLayoutManager
import com.ashlikun.xrecycleview.nested.NestedOnChildTouch
import java.util.*

/**
 * @author　　: 李坤
 * 创建时间: 2021.12.9 16:29
 * 邮箱　　：496546144@qq.com
 *
 * 功能介绍：功能介绍：扩展的RecyclerView
 * 1：带头部与底部的RecycleView
 * 2：上下加载更多
 * 3：和可以设置最大高度
 */

open class RecyclerViewExtend @JvmOverloads constructor(
    context: Context,
    attrs: AttributeSet? = null,
    defStyle: Int = 0
) : RecyclerView(context, attrs, defStyle) {
    companion object {
        const val HEADERSIZE = "headerSize"
        const val FOOTERSIZE = "footerSize"
        private const val TYPE_HEADER = -900004
        private const val TYPE_REFRESH_HEADER = -900005
        private const val TYPE_NORMAL = 0
        private const val TYPE_REFRESH_FOOTER = -900003
        private const val TYPE_FOOTER = -900002
    }

    //头
    var headerViews = ArrayList<View>()

    //尾
    var footViews = ArrayList<View>()

    protected var mAdapter: Adapter<ViewHolder>? = null
    protected var mWrapAdapter: WrapAdapter? = null

    //最大高度
    var maxHeight = 0f
        set(value) {
            field = value
            requestLayout()
        }

    //最大比例,相对于宽度
    var maxRatio = 0f
        set(value) {
            field = value
            requestLayout()
        }

    //是否可以触摸
    var isNoTouch = false

    //是否需要开启和外部的RecyclerView 嵌套滚动
    var nestedOpen = false
        set(value) {
            field = value
            if (field) childTouch = NestedOnChildTouch(this)
        }
    open val dataSize: Int
        get() = mAdapter?.itemCount ?: 0

    //开启和外部的RecyclerView 嵌套滚动 的处理
    protected open lateinit var childTouch: NestedOnChildTouch

    open val headerViewSize: Int
        get() = headerViews.size
    open val footViewSize: Int
        get() = footViews.size
    open val footViewHeight: Int
        get() {
            var height = 0
            footViews.forEach {
                height += it.height
            }
            return height
        }
    open val headerViewHeight: Int
        get() {
            var height = 0
            headerViews.forEach {
                height += it.height
            }
            return height
        }

    init {
        val a = context.obtainStyledAttributes(attrs, R.styleable.RecyclerViewExtend)
        maxRatio = a.getFloat(R.styleable.RecyclerViewExtend_rv_heightRatio, 0f)
        maxHeight = a.getDimension(R.styleable.RecyclerViewExtend_rv_maxHeight, 0f)
        isNoTouch = a.getBoolean(R.styleable.RecyclerViewExtend_rv_noTouch, false)
        nestedOpen = a.getBoolean(R.styleable.RecyclerViewExtend_rv_nested_open, false)
        a.recycle()
        overScrollMode = OVER_SCROLL_NEVER
    }

    fun isHeader(viewHolder: ViewHolder): Boolean {
        return viewHolder.itemViewType == TYPE_REFRESH_HEADER || viewHolder.itemViewType == TYPE_HEADER
    }

    fun isFooter(viewHolder: ViewHolder): Boolean {
        return viewHolder.itemViewType == TYPE_REFRESH_FOOTER || viewHolder.itemViewType == TYPE_FOOTER
    }


    override fun onMeasure(widthMeasureSpec: Int, heightMeasureSpec: Int) {
        if (maxHeight <= 0 && maxRatio <= 0) {
            super.onMeasure(widthMeasureSpec, heightMeasureSpec)
            return
        }
        val widthSize = MeasureSpec.getSize(widthMeasureSpec)
        if (maxHeight <= 0) {
            maxHeight = maxRatio * widthSize
        } else if (maxRatio > 0) {
            maxHeight = Math.min(maxHeight, maxRatio * widthSize)
        }
        if (maxHeight <= 0) {
            super.onMeasure(widthMeasureSpec, heightMeasureSpec)
            return
        }
        val heightMode = MeasureSpec.getMode(heightMeasureSpec)
        var heightSize = MeasureSpec.getSize(heightMeasureSpec)
        if (heightMode == MeasureSpec.EXACTLY) {
            heightSize = if (heightSize <= maxHeight) heightSize else maxHeight.toInt()
        }
        if (heightMode == MeasureSpec.UNSPECIFIED) {
            heightSize = if (heightSize <= maxHeight) heightSize else maxHeight.toInt()
        }
        if (heightMode == MeasureSpec.AT_MOST) {
            heightSize = if (heightSize <= maxHeight) heightSize else maxHeight.toInt()
        }
        val maxHeightMeasureSpec = MeasureSpec.makeMeasureSpec(
            heightSize,
            heightMode
        )
        super.onMeasure(widthMeasureSpec, maxHeightMeasureSpec)
    }

    override fun onTouchEvent(e: MotionEvent): Boolean {
        if (isNoTouch) {
            return false
        }
        return if (nestedOpen && childTouch.onTouchEvent(e)) false else super.onTouchEvent(
            e
        )
    }

    override fun setAdapter(adapter: Adapter<out ViewHolder>?) {
        try {
            mAdapter?.unregisterAdapterDataObserver(mDataObserver)
        } catch (e: IllegalStateException) {
        }
        mAdapter = adapter as Adapter<ViewHolder>?
        mWrapAdapter = WrapAdapter()
        super.setAdapter(mWrapAdapter)
        try {
            mAdapter?.registerAdapterDataObserver(mDataObserver)
        } catch (e: IllegalStateException) {
        }
        mWrapAdapter?.notifyDataSetChanged()
        setHeaderSize()
        setFooterSize()
    }

    override fun getAdapter(): Adapter<out ViewHolder>? {
        return mAdapter
    }

    open fun addHeaderView(view: View) {
        if (view.layoutParams == null) {
            view.layoutParams = LayoutParams(LayoutParams.MATCH_PARENT, LayoutParams.WRAP_CONTENT)
        }
        headerViews.add(view)
        setHeaderSize()
    }

    open fun addHeaderView(index: Int, view: View) {
        if (view.layoutParams == null) {
            view.layoutParams = LayoutParams(LayoutParams.MATCH_PARENT, LayoutParams.WRAP_CONTENT)
        }
        headerViews.add(index, view)
        setHeaderSize()
    }

    open fun putHeaderView(view: View) {
        if (!containsHeaderView(view)) {
            addHeaderView(view)
        }
    }

    open fun putHeaderView(index: Int, view: View) {
        if (!containsHeaderView(view)) {
            addHeaderView(index, view)
        }
    }

    open fun getHeaderView(index: Int) = headerViews.getOrNull(index)

    open fun removeHeaderView(view: View) {
        headerViews.remove(view)
        setHeaderSize()
        mWrapAdapter?.notifyDataSetChanged()
    }

    open fun containsHeaderView(view: View) = headerViews.contains(view)

    open fun removeFootView(view: View) {
        footViews.remove(view)
        setFooterSize()
        mWrapAdapter?.notifyDataSetChanged()
    }

    open fun containsFootView(view: View) = footViews.contains(view)

    open fun addFootView(view: View) {
        if (view.layoutParams == null) {
            view.layoutParams = LayoutParams(LayoutParams.MATCH_PARENT, LayoutParams.WRAP_CONTENT)
        }
        footViews.add(view)
        setFooterSize()
    }

    open fun addFootView(index: Int, view: View) {
        if (view.layoutParams == null) {
            view.layoutParams = LayoutParams(LayoutParams.MATCH_PARENT, LayoutParams.WRAP_CONTENT)
        }
        footViews.add(index, view)
        setFooterSize()
    }

    open fun putFootView(view: View) {
        if (!containsFootView(view)) {
            addFootView(view)
        }
    }

    open fun putFootView(index: Int, view: View) {
        if (!containsFootView(view)) {
            addFootView(index, view)
        }
    }

    open fun getFootView(index: Int) = footViews.getOrNull(index)

    protected open fun setFooterSize() {
        if (mAdapter != null) setHeaderFooterSize(mAdapter!!, FOOTERSIZE, footViewSize)
    }

    protected open fun setHeaderSize() {
        if (mAdapter != null) setHeaderFooterSize(mAdapter!!, HEADERSIZE, headerViewSize)
    }


    open fun scrollToPositionWithOffset(postion: Int) {
        when (val layoutManager = layoutManager) {
            is LinearLayoutManager -> layoutManager.scrollToPositionWithOffset(postion, 0)
            is StaggeredGridLayoutManager -> layoutManager.scrollToPositionWithOffset(postion, 0)
        }
    }


    /**
     * 当adapter不是调用全部刷新，并且有动画的时候
     */
    protected open fun onAdapterItemAnimChang() {}

    private val mDataObserver: AdapterDataObserver = object : AdapterDataObserver() {
        override fun onChanged() {
            mWrapAdapter?.notifyDataSetChanged()
        }

        override fun onItemRangeInserted(positionStart: Int, itemCount: Int) {
            mWrapAdapter?.notifyItemRangeInserted(headerViewSize + positionStart, itemCount)
            if (itemAnimator != null) {
                onAdapterItemAnimChang()
            }
        }

        override fun onItemRangeChanged(positionStart: Int, itemCount: Int) {
            mWrapAdapter?.notifyItemRangeChanged(headerViewSize + positionStart, itemCount)
        }

        override fun onItemRangeChanged(positionStart: Int, itemCount: Int, payload: Any?) {
            mWrapAdapter?.notifyItemRangeChanged(headerViewSize + positionStart, itemCount, payload)
        }

        override fun onItemRangeRemoved(positionStart: Int, itemCount: Int) {
            mWrapAdapter?.notifyItemRangeRemoved(headerViewSize + positionStart, itemCount)
            if (itemAnimator != null) {
                onAdapterItemAnimChang()
            }
        }

        override fun onItemRangeMoved(fromPosition: Int, toPosition: Int, itemCount: Int) {
            mWrapAdapter?.notifyItemMoved(
                headerViewSize + fromPosition,
                headerViewSize + toPosition
            )
            if (itemAnimator != null) {
                onAdapterItemAnimChang()
            }
        }
    }

    protected inner class WrapAdapter : Adapter<ViewHolder>() {
        private var headerPosition = 0
        private var footerPosition = 0
        fun isHeader(position: Int) = position in 0 until headerViews.size
        fun isHeaderLoad(position: Int) =
            headerViews.size > 0 && position == 0 && headerViews[0] is LoadView

        fun isFooter(position: Int): Boolean =
            position in itemCount - footViews.size until itemCount

        fun isFooterLoad(position: Int) =
            footViews.size > 0 && position == itemCount - 1 && footViews[footViews.size - 1] is LoadView

        override fun onAttachedToRecyclerView(recyclerView: RecyclerView) {
            mAdapter?.onAttachedToRecyclerView(recyclerView)
            val manager = recyclerView.layoutManager
            if (manager is GridLayoutManager) {
                manager.spanSizeLookup = SpanSizeLookupGroup(manager, headerViewSize) {
                    isHeader(it) || isFooter(it)
                }
                manager.spanCount = manager.spanCount
            }
        }

        override fun onDetachedFromRecyclerView(recyclerView: RecyclerView) {
            mAdapter?.onDetachedFromRecyclerView(recyclerView)
        }

        override fun onViewAttachedToWindow(holder: ViewHolder) {
            try {
                if (isHeader(holder) || isFooter(holder)) {
                    val lp = holder.itemView.layoutParams
                    if (lp != null && lp is StaggeredGridLayoutManager.LayoutParams
                        && (isHeader(holder.layoutPosition) || isFooter(holder.layoutPosition))
                    ) {
                        //StaggeredGridLayoutManager 只能这样设置一整行
                        lp.isFullSpan = true
                    }
                } else {
                    mAdapter?.onViewAttachedToWindow(holder)
                }
            } catch (e: Exception) {
            }
        }

        override fun onViewDetachedFromWindow(holder: ViewHolder) {
            try {
                if (!isHeader(holder) && !isFooter(holder))
                    mAdapter?.onViewDetachedFromWindow(holder)
            } catch (e: Exception) {
            }
        }

        override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
            if (viewType == TYPE_REFRESH_FOOTER) {
                return object : ViewHolder(getFootView(footViews.size - 1)!!) {}
            } else if (viewType == TYPE_REFRESH_HEADER) {
                return object : ViewHolder(getHeaderView(0)!!) {}
            } else if (viewType == TYPE_HEADER) {
                if (headerPosition >= headerViews.size) {
                    headerPosition = 0
                }
                return object : ViewHolder(headerViews[headerPosition++]) {}
            } else if (viewType == TYPE_FOOTER) {
                if (footerPosition >= footViews.size) {
                    footerPosition = 0
                }
                return object : ViewHolder(footViews[footerPosition++]) {}
            }
            return mAdapter?.onCreateViewHolder(parent, viewType)
                ?: throw RuntimeException("请设置Adapter")
        }

        override fun onBindViewHolder(holder: ViewHolder, position: Int, payloads: List<Any>) {
            super.onBindViewHolder(holder, position, payloads)
            if (isFooterLoad(position) || isHeaderLoad(position)
                || isHeader(position) || isFooter(position)
            ) return
            val adjPosition = position - headerViewSize
            if (mAdapter != null) {
                if (adjPosition < mAdapter?.itemCount ?: 0) {
                    mAdapter?.onBindViewHolder(holder, adjPosition, payloads)
                    return
                }
            }
        }

        override fun onBindViewHolder(holder: ViewHolder, position: Int) {}

        override fun getItemCount() = headerViewSize + footViewSize + (mAdapter?.itemCount ?: 0)

        override fun getItemViewType(position: Int): Int {
            when {
                isFooterLoad(position) -> return TYPE_REFRESH_FOOTER
                isHeaderLoad(position) -> return TYPE_REFRESH_HEADER
                isHeader(position) -> return TYPE_HEADER
                isFooter(position) -> return TYPE_FOOTER
                else -> {
                    val adjPosition = position - headerViewSize
                    val adapterCount = mAdapter?.itemCount ?: 0
                    if (adjPosition < adapterCount) {
                        return mAdapter?.getItemViewType(adjPosition) ?: TYPE_NORMAL
                    }
                    return TYPE_NORMAL
                }
            }
        }

        override fun getItemId(position: Int): Long {
            if (mAdapter != null && position >= headerViewSize) {
                val adjPosition = position - headerViewSize
                val adapterCount = mAdapter!!.itemCount
                if (adjPosition < adapterCount) {
                    return mAdapter!!.getItemId(adjPosition)
                }
            }
            return super.getItemId(position)
        }
    }


}